### DISCUSSION TITLE
vscode: multiple issues with new firejail release

### DISCUSSION THREAD

**Discussion by henrythebuilder** [netblue30/firejail#5027] [Q&A]
Hi,
with my last update for firejail I had to change my local profile for vscode in order to execute the program.

I dont known if all the problems are related to my system/configuration, vscode  or for some firejail profiles, maybe some configurations can only be related to my environment and others could be inserted in global profiles, so this report is to have some help/support to get to a _better configuration_.

My system is `Debian GNU/Linux 11 (bullseye) x86_64` with `firejail version 0.9.68` (`0.9.68-3~bpo11+1`) using vscode self-compiled installed locally in _portable mode_. My old profile/configuration was composed only by 'two' lines only to enable `read-write' operation to vscode local folder.

The problems detected with my local solutions are the following:

## local vscode data folder (obvious modification for my configuration!)
vscode is unable to write logs/data/.. in the local folder, I had to launch firejail/vscode with:
```
firejail --profile=code-oss --read-write=/path/to/local/vscode/installation/data ....
```
whitelisting local folder is not enough. 

## local theme/fonts
vscode is unable to to use my local theme and fonts configuration, GUI is not rendered correctly with local theme/font:
`code-oss.local`
```
whitelist ${HOME}/.themes
read-only ${HOME}/.themes
whitelist ${HOME}/.local/share/fonts
read-only ${HOME}/.local/share/fonts
```

## local source directory
vscode is unable to access to my local source directories (coming from ~/src), always open in _HOME_ folder.
`code-oss.local`
```
whitelist ${HOME}/src
```

## git configuration
vscode is unable to access to global git configuration (~/.gitconfig) for git operations.
`code-oss.local`
```
whitelist ${HOME}/.gitconfig
read-only ${HOME}/.gitconfig
```

## keyboard malfunction
in order to get keyboard shortcut working I had to insert:
`code-oss.local`
```
dbus-user none
dbus-system none
```  

I attach the complete configuration.

Thanks in advance
Enrico

------
Firejail command line:
```
firejail --profile=code-oss --read-write=/path/to/local/vscode/installation/data /path/to/vscode/code-oss ...
```
`code-oss.local` profile:
```
# Error: no suitable /.../code-oss executable found
whitelist ${HOME}/bin/VSCode-linux-x64

# local user themes/fonts/icon
whitelist ${HOME}/.themes
read-only ${HOME}/.themes
whitelist ${HOME}/.local/share/fonts
read-only ${HOME}/.local/share/fonts

# extra dev
whitelist ${HOME}/.gitconfig
read-only ${HOME}/.gitconfig
whitelist ${HOME}/src

# https://github.com/netblue30/firejail/issues/3618#issuecomment-714423674
# OK keyboard!!!
dbus-user none
dbus-system none
```

**Comment by rusty-snake** [comment]
vscode? Did you check #5012 and #5019?

**Reply by henrythebuilder** [reply]
Yes, I checked: 
about #5012 my problem is not related to 'electron' (I suppose), I don't have any `${HOME}/.config/Electron` or `${HOME}/.config/electron-flag*.conf`. vscode was unable to open my source files in `~/src/...` sub-directories, always open the 'HOME' folder without a specific configuration for `~/src` folder (I don't want to whitelist the entire home directory).
For #5019 I had no problem with configuration, also with older firejail, given that I have always had to give read/write permissions to the folder of vscode settings/data 
I don't know if may be related to this two case but I always use vscode from command line, so for firejail I use the script in the `bin` folder (`vscode-installation/bin/code-oss`) instead the main executable (`vscode-installation/code-oss`)

**Reply by rusty-snake** [reply]
So your actual question is ... ?

**Reply by rusty-snake** [reply]
That you need `dbus-user none` is maybe related to ibus/fcitix if you use. Do you?

**Reply by henrythebuilder** [reply]
_"So your actual question is ... ?"_
The thing that led me to write the message is that it seemed strange that actual profiles did not give the application access practically to nothing, especially for common configurations such as fonts and local themes compared to that used previously in the same system.
This is probably also because I didn't follow the changes for new versions in details, especially for new profiles, and even perhaps for  _vscode internal_ changes.
Given this the profile created seems to work, so it's okay, thanks for answering and for work on Firejail.
I apologize for the inconvenience.

**Reply by rusty-snake** [reply]
> The thing that led me to write the message is that it seemed strange that actual profiles did not give the application access practically to nothing, especially for common configurations such as fonts and local themes compared to that used previously in the same system.

That is #5012/#5019.
And fixed by #5026.

**Reply by henrythebuilder** [reply]
I tried to update only the new `code.profile` (from #5026) but  no success without local change (no 'interface' with local theme/font and access to source folder), however I wait for the next complete update of the package to check the tests again :+1:

**Reply by rusty-snake** [reply]
Did you commented you code.local changes for testing?

**Reply by henrythebuilder** [reply]
Yes, I left only the whitelist for the vscode local folder root (portable installation)

**Reply by henrythebuilder** [reply]
... I also tried to launch bash with the last code profile:
`firejail --profile=code-oss /bin/bash`
```
$ ls -la ~
totale 12
drwx------ 3 user user  120 11 mar 08.39 .
drwxr-xr-x 3 root root   60 11 mar 08.39 ..
-rw-r--r-- 1 user user 3526 11 mar 08.39 .bashrc
drwxr-xr-x 3 user user   60 11 mar 08.39 bin
-rw-r--r-- 1 user user   26 11 mar 08.39 .inputrc
-rw------- 1 user user   59 11 mar 08.39 .Xauthority
```
 `bin` folder is due to the whitelist!

**Reply by henrythebuilder** [reply]
... apart from the particular configurations of my installation, probably there are no references to the system folders needed for the configuration:
`firejail --profile=chromium /bin/bash`
...
```
$ ls -la ~
totale 36
drwx------ 10 user user  280 11 mar 09.00 .
drwxr-xr-x  3 root root   60 11 mar 09.00 ..
-rw-r--r--  1 user user 3526 11 mar 09.00 .bashrc
drwxr-xr-x  2 user user   40 11 mar 09.00 .cache
drwxr-xr-x 11 user user  300 11 mar 09.00 .config
drwxr-xr-x 28 user user 4096  8 nov 17.35 Downloads
drwx------  6 user user 4096 16 ago  2021 .gnome2
-rw-r-----  1 user user  245  4 gen 10.27 .gtkrc-2.0
drwxr-x---  2 user user 4096 25 ago  2021 .icons
-rw-r--r--  1 user user   26 11 mar 09.00 .inputrc
drwxr-xr-x  3 user user   60 11 mar 09.00 .local
drwx------  3 user user 4096 15 nov  2015 .pki
drwxr-x--- 20 user user 4096  1 set  2021 .themes
-rw-------  1 user user   59 11 mar 09.00 .Xauthority
```

**Reply by henrythebuilder** [reply]
sorry for the previous message, the atom profile is more suitable perhaps because it includes electron (as vscode), the result is the same:
`firejail --profile=atom /bin/bash`

```
$  ls -la ~
totale 12
drwx------ 2 user user  100 11 mar 09.37 .
drwxr-xr-x 3 root root   60 11 mar 09.37 ..
-rw-r--r-- 1 user user 3526 11 mar 09.37 .bashrc
-rw-r--r-- 1 user user   26 11 mar 09.37 .inputrc
-rw------- 1 user user   59 11 mar 09.37 .Xauthority
```

**Comment by henrythebuilder** [comment]
## Testing update:
Adding the content of  `whitelist-common.inc` to my `code-oss.local` problems related to _"GUI theme/font"_ and _"keyboard malfunction"_ have been solved, there is no need for _dbus_ directive and/or _whitelist 'theme/font'_ folders!
In this way the remaining local configuration is related only to my local installation.
It remains to understand which is the minimum content of the `whitelist-common.inc` to be included (and/or modified) for development environments in order to ensure the best safety.

**Reply by rusty-snake** [reply]
I wouldn't mind minimizing wc but if you like you can.

**Reply by henrythebuilder** [reply]
At the moment I left the  content of `whitelist-common.inc` as it is for my local configuration.
For me the important thing was to solve the problem using how much already exists in the Firejail profiles!
Thanks again for the  support and all the contributors of the project!
